kube-runtime/src/watcher.rs

```diff
@@ -138,48 +150,89 @@ async fn step_trampolined<K: Resource + Clone + DeserializeOwned + Debug + Send
     state: State<K>,
 ) -> (Option<Result<Event<K>>>, State<K>) {
     match state {
-        State::Empty => match api.list(list_params).await {
-            Ok(list) => (Some(Ok(Event::Restarted(list.items))), State::InitListed {
-                resource_version: list.metadata.resource_version.unwrap(),
-            }),
-            Err(err) => (Some(Err(err).map_err(Error::InitialListFailed)), State::Empty),
-        },
-        State::InitListed { resource_version } => match api.watch(list_params, &resource_version).await {
-            Ok(stream) => (None, State::Watching {
-                resource_version,
-                stream: stream.boxed(),
-            }),
-            Err(err) => (
-                Some(Err(err).map_err(Error::WatchStartFailed)),
-                State::InitListed { resource_version },
-            ),
-        },
+        State::Empty {
+            continue_token,
+            mut objects,
+        } => {
+            let mut lp = list_params.clone();
+            lp.continue_token = continue_token;
+            lp.limit = Some(100);
+            match api.list(&lp).await {
+                Ok(list) => {
+                    objects.extend(list.items);
+                    if let Some(continue_token) = list.metadata.continue_ {
+                        (
+                            None,
+                            State::Empty {
+                                continue_token: Some(continue_token),
+                                objects,
+                            },
+                        )
+                    } else {
+                        (
+                            Some(Ok(Event::Restarted(objects))),
+                            State::InitListed {
+                                resource_version: list.metadata.resource_version.unwrap(),
+                            },
+                        )
+                    }
+                }
```

Comment:
I like that this is basically the core of pagination. Very clean if it works. Probably needs a similar thing in the metadata variant, but huge +1 on this part from me.

Reply:
I believe this should automatically carry over now that we use `ApiMode`, right?

Reply:
oh, yep, you are right. forgot we deduplicated them.